fdbserver/GlobalTagThrottler.actor.cpp

```diff
@@ -0,0 +1,191 @@
+				} else {
+					totalLimit = std::max(totalLimit, std::min(readLimit.get(), writeLimit.get()));
+				}
+				return ClientTagThrottleLimits(totalLimit, std::numeric_limits<double>::max());
+			}
+		}
+	};
+
+	Database db;
+	UID id;
+	std::map<TransactionTag, QuotaAndCounters> trackedTags;
+	uint64_t throttledTagChangeId{ 0 };
+
+	ACTOR static Future<Void> monitorThrottlingChanges(GlobalTagThrottlerImpl* self) {
+		loop {
+			state ReadYourWritesTransaction tr(self->db);
+
+			loop {
+				// TODO: Clean up quotas that have been removed
+				try {
+					tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+					tr.setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
+
+					state RangeResult currentQuotas = wait(tr.getRange(tagQuotaKeys, CLIENT_KNOBS->TOO_MANY));
+					TraceEvent("GlobalTagThrottler_ReadCurrentQuotas").detail("Size", currentQuotas.size());
+					for (auto const kv : currentQuotas) {
+						auto tag = kv.key.removePrefix(tagQuotaPrefix);
+						auto quota = ThrottleApi::TagQuotaValue::fromValue(kv.value);
+						self->trackedTags[tag].setQuota(quota);
+					}
+
+					++self->throttledTagChangeId;
+					// FIXME: Should wait on watch instead
+					// wait(tr.watch(tagThrottleSignalKey));
+					wait(delay(5.0));
+					TraceEvent("GlobalTagThrottler_ChangeSignaled");
+					TEST(true); // Global tag throttler detected quota changes
+					break;
+				} catch (Error& e) {
+					TraceEvent("GlobalTagThrottlerMonitoringChangesError", self->id).error(e);
+					wait(tr.onError(e));
+				}
+			}
+		}
+	}
+
+public:
+	GlobalTagThrottlerImpl(Database db, UID id) : db(db), id(id) {}
+	Future<Void> monitorThrottlingChanges() { return monitorThrottlingChanges(this); }
+	void addRequests(TransactionTag tag, int count) { trackedTags[tag].addTransactions(count); }
+	uint64_t getThrottledTagChangeId() const { return throttledTagChangeId; }
+	PrioritizedTransactionTagMap<ClientTagThrottleLimits> getClientRates() {
+		// TODO: For now, only enforce total throttling rates.
+		// We should use reserved quotas as well.
+		PrioritizedTransactionTagMap<ClientTagThrottleLimits> result;
+		for (const auto& [tag, quotaAndCounters] : trackedTags) {
+			// Currently there is no differentiation between batch priority and default priority transactions
+			auto const limit = quotaAndCounters.getTotalLimit();
+			if (limit.present()) {
+				result[TransactionPriority::BATCH][tag] = result[TransactionPriority::DEFAULT][tag] = limit.get();
+			}
+		}
+		return result;
+	}
+	int64_t autoThrottleCount() const { return trackedTags.size(); }
+	uint32_t busyReadTagCount() const {
+		// TODO: Implement
+		return 0;
+	}
+	uint32_t busyWriteTagCount() const {
+		// TODO: Implement
+		return 0;
+	}
+	int64_t manualThrottleCount() const { return trackedTags.size(); }
+	Future<Void> tryUpdateAutoThrottling(StorageQueueInfo const& ss) {
+		for (const auto& busyReadTag : ss.busiestReadTags) {
+			trackedTags[busyReadTag.tag].addReadCost(busyReadTag.rate);
```

Comment:
Does using the N busiest tags pose a problem if some tags have a much lower quota than others?

Reply:
You're right, if some quotas are extremely small, this is a potential problem, and would motivate making `N` large.

Reply:
It seems like the ratekeeper should have a way to get the sum of read/write cost of every tag to have a global view instead of the sum of the topK tags reported by storage servers.